sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/CachingBeamFnStateClient.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.cache.LoadingCache;
+
+/**
+ * Wraps a delegate BeamFnStateClient and stores the result of state requests in cross bundle cache
+ * according to the available cache tokens. If there are no cache tokens for the state key requested
+ * the request is forwarded to the client and executed normally.
+ */
+public class CachingBeamFnStateClient implements BeamFnStateClient {
+
+  private final BeamFnStateClient beamFnStateClient;
+  private final LoadingCache<StateKey, Map<StateCacheKey, StateGetResponse>> stateCache;
+  private final Map<CacheToken.SideInput, ByteString> sideInputCacheTokens;
+  private final ByteString userStateToken;
+
+  /**
+   * Creates a CachingBeamFnStateClient that wraps a BeamFnStateClient with a LoadingCache. Cache
+   * tokens are sent by the runner to indicate which state is able to be cached.
+   */
+  public CachingBeamFnStateClient(
+      BeamFnStateClient beamFnStateClient,
+      LoadingCache<StateKey, Map<StateCacheKey, StateGetResponse>> stateCache,
+      List<CacheToken> cacheTokenList) {
+    this.beamFnStateClient = beamFnStateClient;
+    this.stateCache = stateCache;
+    this.sideInputCacheTokens = new HashMap<>();
+
+    // Set up cache tokens.
+    ByteString tempUserStateToken = ByteString.EMPTY;
+    for (BeamFnApi.ProcessBundleRequest.CacheToken token : cacheTokenList) {
+      if (token.hasUserState()) {
+        tempUserStateToken = token.getToken();
+      } else if (token.hasSideInput()) {
+        sideInputCacheTokens.put(token.getSideInput(), token.getToken());
+      }
+    }
+    this.userStateToken = tempUserStateToken;
+  }
+
+  /**
+   * Completes the response with a cached value if possible, if not forwards the response to the
+   * BeamFnStateClient and tries caching the result. All Append and Clear requests are forwarded.
+   */
+  @Override
+  public void handle(
+      StateRequest.Builder requestBuilder, CompletableFuture<StateResponse> response) {
+
+    StateKey stateKey = requestBuilder.getStateKey();
+    ByteString cacheToken = getCacheToken(stateKey);
+
+    // If state is not cacheable proceed as normal.
+    if (ByteString.EMPTY.equals(cacheToken)) {
+      beamFnStateClient.handle(requestBuilder, response);
+      return;
+    }
+
+    switch (requestBuilder.getRequestCase()) {
+      case GET:
+        // Check if data is in the cache.
+        StateCacheKey cacheKey =
+            StateCacheKey.create(cacheToken, requestBuilder.getGet().getContinuationToken());
+        Map<StateCacheKey, StateGetResponse> stateKeyMap = stateCache.getUnchecked(stateKey);
+        StateGetResponse cachedPage = stateKeyMap.get(cacheKey);
+
+        // If data is not cached, add callback to add response to cache on completion.
+        // Otherwise, complete the response with the cached data.
+        if (cachedPage == null) {
+          beamFnStateClient.handle(requestBuilder, response);
```

Comment:
Question: how is the GET served when cache miss? When cache hits, you have `response.complete`, do you need the same for cache miss after you put the state into the cache?

Reply:
The response.complete is executed by the beamFnStateClient.handle call here, but if it is a cache hit we can complete the response immediately so we do not need to forward the request to the delegate state client

Reply:
Then can you confirm that after complete, your thenAccept will still be executed? (I am not a async API expert so I am not sure).